packages/app/auth/providers/earthdata-login.ts

```diff
@@ -0,0 +1,115 @@
+import type { OAuthConfig, OAuthUserConfig } from 'next-auth/providers/oauth'
+import type { TokenSetParameters } from 'openid-client'
+
+export interface EarthdataUser {
+    uid: string
+    first_name: string
+    last_name: string
+    email_address: string
+    registered_date: string
+    country: string
+    study_area: string
+    allow_auth_app_emails: boolean
+    user_type: string
+    affiliation: string
+    agreed_to_meris_eula: boolean
+    agreed_to_sentinel_eula: boolean
+    email_verified: boolean
+    user_authorized_apps: number
+}
+
+export const basePath = 'https://urs.earthdata.nasa.gov'
+
+export type EDLTokenSetParameters = Pick<
+    TokenSetParameters,
+    'access_token' | 'token_type' | 'refresh_token'
+> & {
+    expires_in: number
+    endpoint: string
+}
+/**
+ * A custom OAuth2 Provider for authenticating to Earthdata Login
+ * (https://urs.earthdata.nasa.gov, https://next-auth.js.org/configuration/providers/oauth)
+ */
+export default function EarthdataLoginProvider<P extends EarthdataUser = any>(
+    options: OAuthUserConfig<P>
+): OAuthConfig<P> {
+    return {
+        id: 'earthdata-login',
+        name: 'Earthdata Login',
+        version: '2.0',
+        type: 'oauth',
+        authorization: {
+            // https://urs.earthdata.nasa.gov/documentation/for_integrators/api_documentation#GET/oauth/authorize
+            url: `${basePath}/oauth/authorize`,
+            params: {
+                // disables the URS "splash screen" (the one that says "Redirecting...") for logged in users
+                splash: false,
+            },
+        },
+        token: {
+            /**
+             * Earthdata's OAuth2 token retrieval deviates from the Next Auth specification
+             * so we will send out our own POST token request
+             */
+            async request({ params, provider }) {
```

Comment:
This is where it gets a little hairy, the token endpoint differs from a normal oAuth implementation.

We're basically just triggering our own fetch rather than using NextAuths built-in (this isn't a hack, it's a normal way of doing custom providers)